File: algorithms/C++/Arrays/remove-duplicates.cpp
/*CPP program to remove the duplicate elements from the array in place
  without using any extra space.
  
  The time complexity of the algorithm is: O(n)
Sample Example:
Input array size: 6
Input array: 1 1 2 3 4 4
output : 1 2 3 4
*/

#include <bits/stdc++.h>
using namespace std;

// method to remove duplicate elements from array
void removeDuplicates(int arr[],int size){
	 
int newsize=size; // variable to keep the track of size of array.

for(int i=0,j=i+1;j<size;j++){

     if(arr[i]==arr[j]){
            newsize--;
        }else{
               i++;
    	       arr[i]=arr[j];
              }

        }
  // displaying the array after removing the duplicate elements.
    cout<<"The Array after removing duplicate elements are:"<<endl;
        for(int i=0;i<newsize;i++){
               cout<<arr[i]<<endl;
            }

   }


// driver program to test above methods
int main(){
	// get the size of the array
   int size;
   cout<<"Enter the size of the array"<<endl;
   cin>>size;
   // declare the array
   int arr[size];
  // storing the minimum integer value for checking of array is sorted or not
int max=-2147483647;

   // get the  sorted array from the user
   cout<<"Enter the sorted array"<<endl;
 
   for(int i=0;i<size;i++){
   	cin>>arr[i];
   	// checking if input array is sorted or not
   	if(arr[i]<max){
   		cout<<"Array not sorted";
   		return 0;
	   }else{
	   	max=arr[i];
	   }
   }
    removeDuplicates(arr,size);

}

File: algorithms/Go/arrays/remove-duplicates.go
//Description: Given an integer array nums sorted in non-decreasing order, remove the duplicates in-place such that each unique ///element appears only once. The relative order of the elements should be kept the same. Then return the number of unique elements in nums
//Time Complexity: O(n)
//Space Complexity: O(n)

//Input: nums = [1,1,2]
//Output: 2, nums = [1,2,_]

package main

import "fmt"

func removeDuplicates(nums []int) int {
	ans := []int{}
	var j int = 0

	ans = append(ans, nums[0])

	for i := 0; i < len(nums); i++ {
		if nums[i] != nums[j] {
			j = i
			ans = append(ans, nums[i])
		}
	}
	// for i := 0; i < len(ans); i++ {
	// 	nums[i] = ans[i]
	// }
	copy(nums, ans)
	_ = j
	return len(ans)
}

func mainRemoveDuplicates() {
	arr := []int{1, 2, 2, 3, 4, 4, 4, 5}
	res := removeDuplicates(arr)
	fmt.Println(res)
}
